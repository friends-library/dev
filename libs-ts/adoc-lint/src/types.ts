import { Lang, EditionType } from '@friends-library/types';

export interface LintResult {
  line: number;
  column: number | false;
  type: 'error' | 'warning' | 'notice';
  rule: string;
  message: string;
  recommendation?: string;
  fixable?: boolean;
  info?: string;
}

export interface LintOptions {
  lang: Lang;
  editionType?: EditionType;
  include?: string[];
  exclude?: string[];
  maybe?: boolean;
}

export interface LineRule {
  (line: string, lines: string[], lineNumber: number, options: LintOptions): LintResult[];
  slug: string;
  maybe?: boolean;
}

export interface BlockRule {
  (block: string, options: LintOptions): LintResult[];
  slug: string;
  maybe?: boolean;
}
