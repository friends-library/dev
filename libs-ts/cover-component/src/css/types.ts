export interface CoverCssModule {
  (scaler?: number, scope?: string): [string, string];
}
