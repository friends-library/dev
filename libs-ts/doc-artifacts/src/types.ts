export interface FileManifest {
  [key: string]: string | Buffer;
}

export interface Options {
  namespace?: string;
  srcPath?: string;
}

export type PdfOptions = Options & {
  formatOutput?: (line: string) => string;
};

export type EbookOptions = Options & {
  check: boolean;
};
