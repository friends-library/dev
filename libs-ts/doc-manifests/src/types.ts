import { PrintSize } from '@friends-library/types';

export interface PaperbackInteriorConfig {
  printSize: PrintSize;
  frontmatter: boolean;
  condense: boolean;
  allowSplits: boolean;
}

export interface PaperbackCoverConfig {
  printSize: PrintSize;
  volumes: number[];
  showGuides?: boolean;
}

export interface EbookConfig {
  frontmatter: boolean;
  subType: 'epub' | 'mobi';
  coverImg?: Buffer;
  randomizeForLocalTesting?: boolean;
}
