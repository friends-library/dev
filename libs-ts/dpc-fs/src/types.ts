import { DocPrecursor } from '@friends-library/types';

export type FsDocPrecursor = DocPrecursor & { fullPath: string };
