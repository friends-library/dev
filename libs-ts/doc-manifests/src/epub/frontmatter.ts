import { DocPrecursor, FileManifest, Lang } from '@friends-library/types';
import { ChapterResult, EbookSrcResult } from '@friends-library/evaluator';
import { frontmatter as commonFrontmatter } from '../frontmatter';

export default function frontmatter(
  dpc: DocPrecursor,
  src: EbookSrcResult,
  target: 'mobi' | 'epub',
): FileManifest {
  const fm = commonFrontmatter(dpc, src);
  fm[`half-title`] = `<div class="half-title-page">${fm[`half-title`]}</div>`;

  if (src.hasFootnotes) {
    fm[`footnote-helper`] = src.footnoteHelperSourceHtml;
  }

  if (target === `mobi` && src.numChapters > 2) {
    fm[`content-toc`] = contentToc(src.chapters, dpc.lang);
  }

  return fm;
}

function contentToc(chapters: ChapterResult[], lang: Lang): string {
  return `
  <section class="content-toc">
    <h1>${lang === `en` ? `Table of Contents` : `√çndice`}</h1>
    ${chapters.map(tocEntry).join(`\n`)}
  </section>`;
}

function tocEntry(chapter: ChapterResult): string {
  const short = chapter.shortHeading;
  return `
  <div>
    <a href="${chapter.id}.xhtml">
      ${chapter.isIntermediateTitle ? `~ ${short} ~` : short}
    </a>
  </div>`;
}
