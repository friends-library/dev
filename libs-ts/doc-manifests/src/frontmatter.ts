import moment from 'moment';
import { pickBy } from 'lodash';
import { HTML_DEC_ENTITIES, DocPrecursor, FileManifest } from '@friends-library/types';
import { htmlTitle } from '@friends-library/adoc-utils';
import { HtmlSrcResult } from '@friends-library/evaluator';
import { t, setLocale } from '@friends-library/locale';
import { capitalizeTitle, ucfirst } from './utils';
import { addVolumeSuffix } from './faux-volumes';

export function frontmatter(dpc: DocPrecursor, src: HtmlSrcResult): FileManifest {
  const files = {
    'half-title': halfTitle(dpc),
    'original-title': originalTitle(dpc),
    copyright: copyright(dpc),
    epigraph: src.epigraphHtml,
  };
  return pickBy(files, (html) => html !== ``);
}

export function halfTitle(dpc: DocPrecursor, volIdx?: number): string {
  const {
    lang,
    meta: {
      title,
      editor,
      author: { name },
    },
  } = dpc;

  const prettyTitle = htmlTitle(addVolumeSuffix(title, volIdx)).replace(
    name,
    name.replace(/ /g, HTML_DEC_ENTITIES.NON_BREAKING_SPACE),
  );

  let markup = `<h1>${prettyTitle}</h1>`;
  const nameInTitle = title.indexOf(name) !== -1;
  if (!nameInTitle && !dpc.isCompilation) {
    markup = `${markup}\n<p class="byline">${lang === `en` ? `by` : `por`} ${name}</p>`;
  }

  if (editor && lang === `en`) {
    markup += `\n<p class="editor">Edited by ${editor}</p>`;
  }

  return markup;
}

export function originalTitle({ meta, lang }: DocPrecursor): string {
  if (!meta.originalTitle) {
    return ``;
  }

  setLocale(lang);
  return `
    <div class="original-title-page">
      <p class="originally-titled__label">
        ${t`Original title`}:
      </p>
      <p class="originally-titled__title">
        ${capitalizeTitle(meta.originalTitle, lang)}
      </p>
    </div>
  `;
}

export function copyright(dpc: DocPrecursor): string {
  const {
    lang,
    revision: { timestamp, sha, url },
    meta: { published, isbn },
  } = dpc;
  setLocale(lang);
  moment.locale(lang);

  let time = moment
    .utc(moment.unix(timestamp))
    .format(lang === `en` ? `MMMM Do, YYYY` : `D [de] MMMM, YYYY`);

  if (lang === `es`) {
    time = time
      .split(` `)
      .map((p) => (p === `de` ? p : ucfirst(p)))
      .join(` `);
  }

  const webUrl = `https://www.${t`friendslibrary.com`}`;
  const email = t`info@friendslibrary.com`;

  return `
  <div class="copyright-page">
    <ul>
      <li>${t`Public domain in the USA`}</li>
      ${published ? `<li>${t`Originally published in ${published}`}</li>` : ``}
      ${isbn ? `<li id="isbn">ISBN: <code>${isbn}</code></li>` : ``}
      <li>${t`Text revision ${`<code><a href="${url}">${sha}</a></code> â€” ${time}`}`}</li>
      <li>${t`Created and freely distributed by`} <a href="${webUrl}">${t`Friends Library Publishing`}</a></li>
      <li>${t`Find more free books from early Quakers at`} <a href="${webUrl}">${t`friendslibrary.com`}</a></li>
      <li>${t`Contact the publishers at`} <a href="mailto:${email}">${email}</a></li>
    </ul>
  </div>
  `;
}
