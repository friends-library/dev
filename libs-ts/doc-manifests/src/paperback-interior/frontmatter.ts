import { toRoman } from 'roman-numerals';
import {
  originalTitle as commonOriginalTitle,
  copyright as commonCopyright,
  halfTitle as commonHalfTitle,
} from '../frontmatter';
import { DocPrecursor } from '@friends-library/types';
import { PdfSrcResult, ChapterResult } from '@friends-library/evaluator';
import { rangeFromVolIdx } from '../faux-volumes';

export default function frontmatter(
  dpc: DocPrecursor,
  src: PdfSrcResult,
  volIdx?: number,
): string {
  const isFirstOrOnlyVolume = typeof volIdx !== `number` || volIdx === 0;
  return `
    ${halfTitle(dpc, volIdx)}
    ${isFirstOrOnlyVolume ? originalTitle(dpc) : ``}
    ${copyright(dpc)}
    ${isFirstOrOnlyVolume ? src.epigraphHtml : ``}
    ${toc(src, dpc, volIdx)}
  `;
}

function toc(src: PdfSrcResult, dpc: DocPrecursor, volIdx?: number): string {
  if (src.numChapters <= 3) {
    return ``;
  }
  const toEntry = useMultiColLayout(src.chapters) ? multiColTocEntry : tocEntry;
  return `
    <div class="toc own-page">
      <h1>${dpc.lang === `en` ? `Contents` : `√çndice`}</h1>
      ${src.chapters
        .slice(...rangeFromVolIdx(dpc.paperbackSplits, volIdx))
        .map(toEntry)
        .join(`\n`)}
    </div>
  `;
}

export function useMultiColLayout(chapters: ChapterResult[]): boolean {
  // keep only chapters with headings like `"Chapter IV. The Lambs War"`
  // which, when they predominate, make the multi-col layout desirable
  const numberedAndNamedChapters = chapters.filter(
    (c) => !c.isIntermediateTitle && c.isSequenced && c.hasNonSequenceTitle,
  );
  const nonItermediateChapters = chapters.filter((c) => !c.isIntermediateTitle);
  return numberedAndNamedChapters.length / nonItermediateChapters.length > 0.45;
}

function multiColTocEntry(chapter: ChapterResult): string {
  if (chapter.isIntermediateTitle) {
    return tocEntry(chapter);
  }

  // if we have a sequence (chapter) number, and a non-sequence title
  // we split the chapter number left, and non-sequence title right
  // otherwise, everything goes on right
  const splittable =
    typeof chapter.sequenceNumber === `number` && chapter.hasNonSequenceTitle;

  let main = chapter.shortHeading;
  if (chapter.nonSequenceTitle && chapter.nonSequenceTitle.length < main.length) {
    main = chapter.nonSequenceTitle;
  }

  return `
    <p class="multicol-toc-entry">
      <a href="#${chapter.id}">
        <span class="multicol-toc-chapter">
          ${splittable ? toRoman(chapter.sequenceNumber ?? 1) : ``}
        </span>
        <span class="multicol-toc-main">
          ${main}
        </span>
      </a>
    </p>
    `.trim();
}

function tocEntry(chapter: ChapterResult): string {
  return `
    <p${chapter.isIntermediateTitle ? ` class="toc-intermediate-title"` : ``}>
      <a href="#${chapter.id}">
        <span>${chapter.shortHeading}</span>
      </a>
    </p>`;
}

function copyright(dpc: DocPrecursor): string {
  return commonCopyright(dpc)
    .replace(`copyright-page`, `copyright-page own-page`)
    .replace(`Ebook created`, `Created`)
    .replace(/([^@])friendslibrary\.com/g, `$1www.friendslibrary.com`);
}

function halfTitle(dpc: DocPrecursor, volIdx?: number): string {
  return `
    <div class="half-title-page own-page">
      <div>
        ${commonHalfTitle(dpc, volIdx)}
      </div>
    </div>
  `;
}

function originalTitle(dpc: DocPrecursor): string {
  if (!dpc.meta.originalTitle) {
    return ``;
  }

  return `
    <div class="blank-page own-page"></div>
    ${commonOriginalTitle(dpc)}
  `;
}
