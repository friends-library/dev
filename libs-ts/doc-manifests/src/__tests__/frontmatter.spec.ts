import { describe, test, expect } from '@jest/globals';
import { genericDpc } from '@friends-library/types';
import { copyright } from '../frontmatter';

describe(`copyright()`, () => {
  test(`english copyright`, () => {
    expect(copyright(genericDpc())).toMatchSnapshot();
  });

  test(`spanish copyright`, () => {
    const dpc = genericDpc();
    dpc.lang = `es`;
    expect(copyright(dpc)).toMatchSnapshot();
  });
});
