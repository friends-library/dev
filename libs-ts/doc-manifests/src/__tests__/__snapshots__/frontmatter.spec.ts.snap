// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`copyright() english copyright 1`] = `
"
  <div class=\\"copyright-page\\">
    <ul>
      <li>Public domain in the USA</li>
      
      <li id=\\"isbn\\">ISBN: <code>978-1-64476-029-1</code></li>
      <li>Text revision <code><a href=\\"\\">327ceb2</a></code> — January 22nd, 2021</li>
      <li>Created and freely distributed by <a href=\\"https://www.friendslibrary.com\\">Friends Library Publishing</a></li>
      <li>Find more free books from early Quakers at <a href=\\"https://www.friendslibrary.com\\">friendslibrary.com</a></li>
      <li>Contact the publishers at <a href=\\"mailto:info@friendslibrary.com\\">info@friendslibrary.com</a></li>
    </ul>
  </div>
  "
`;

exports[`copyright() spanish copyright 1`] = `
"
  <div class=\\"copyright-page\\">
    <ul>
      <li>Dominio público en los Estados Unidos de América</li>
      
      <li id=\\"isbn\\">ISBN: <code>978-1-64476-029-1</code></li>
      <li>Revisión de texto <code><a href=\\"\\">327ceb2</a></code> — 22 de Enero, 2021</li>
      <li>Creado y distribuido gratuitamente por <a href=\\"https://www.bibliotecadelosamigos.org\\">La Biblioteca de los Amigos</a></li>
      <li>Encuentre más libros gratis de los primeros Cuáqueros en <a href=\\"https://www.bibliotecadelosamigos.org\\">bibliotecadelosamigos.org</a></li>
      <li>Puede contactarnos en <a href=\\"mailto:info@bibliotecadelosamigos.org\\">info@bibliotecadelosamigos.org</a></li>
    </ul>
  </div>
  "
`;
